dataplicity/portforward.py

```diff
@@ -320,12 +324,34 @@ def open(self, m2m_port, service=None, port=None):
         service.connect(m2m_port)
 
     def redirect_port(self, m2m_port, device_port):
+        # we need to store the reference to the Service somewhere so that
+        # when the Connection starts in thread it wouldn't loose the value
+        # of service variable. However, we have to remember that there may
+        # be numerous connections to the same local port.
+        # for instance, one could be ssh'ed into a machine twice, so we
+        # shan't confuse these two connections.
+        # therefore, an easy way is to store these in a dict, so that the
+        # lookup would be quick
+        #
+        # if there are no redirects on this port, create a hash
         if device_port not in self._dynamic_ports:
-            service = Service(
-                manager=self, name='port-{}'.format(device_port),
-                port=device_port, host='127.0.0.1'
-            )
-            self._dynamic_ports[device_port] = service
-        service = self._dynamic_ports[device_port]
+            self._dynamic_ports[device_port] = {}
+        # create the service which will handle traffic trough Connection
+        service = Service(
+            manager=self, name='port-{}'.format(device_port),
+            port=device_port, host='127.0.0.1'
+        )
+        # store the reference, so that it doesn't deref in Connection thread
+        self._dynamic_ports[device_port][m2m_port] = service
+        # connect to M2M
         service.connect(m2m_port)
+        # the below line is merely a shortcut for nicer-looking unit-tests.
         return service
+
+    def remove_service(self, m2m_port, device_port):
+        # great. The connection was closed, so we need to clean up. However
+        # this function would also be executed after a forward to port 80
+        # (which is noto dynamically redirected) would finish.
+        if device_port in self._dynamic_ports:
```

Comment:
that is why there is this `if` statement. Wormhole uses 'regular' port redirects, i.e. `self._services` and `self._ports`. That's why I am checking this - the only way for `device_port` to be present in `_dynamic_ports` is when somebody uses dynamic redirect.

Whether there is a static redirect (created in the constructor) as well is irrelevant if I am not mistaken - you could have two redirects (one from the constructor, the other one from dynamic redirect)